### PR TITLE
colexec: allow unneeded index columns to be skipped in cfetcher

### PR DESCRIPTION
**Pull request by rohany**
It seems like all indexed columns were being marked as needed in
the cfetcher initialization steps, whether they were actually
in neededCols or not. This meant that unneeded columns weren't
actually getting skipped.

Release note: None

### CODE REVIEW

**Comment by rohany** [comment]
I'm not sure if this is correct, so running the test suite to see if my hunch is right. I noticed this by wondering why https://github.com/cockroachdb/cockroach/pull/43002 doesn't fix https://github.com/cockroachdb/cockroach/issues/42994. In the normal row fetcher all indexed columns are decoded, but the cfetcher has the ability to skip keys.

**Comment by rohany** [comment]
This depends on https://github.com/cockroachdb/cockroach/pull/43002 after the recent push.

**Review by jordanlewis — COMMENTED** [review]
This needs some tests that prove that what you say is being added is actually true. Like, make a query that doesn't read all index columns and have some unsupported types in there.

It looks like it's failing CI due to an unsupported type thing?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43024#-:-LwGLliD1WQRXpKlycLX:b6ihwqt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis)

File: pkg/sql/colexec/cfetcher.go [hunk 1]
```diff
@@ -69,6 +69,9 @@ type cTableInfo struct {
 	// One value per column that is part of the key; each value is a column
 	// index (into cols); -1 if we don't need the value for that column.
 	indexColOrdinals []int
+	// allIndexColOrdinals is the same as allIndexColOrdinals but
```
Comment (jordanlewis) [inline pkg/sql/colexec/cfetcher.go]: This sentence references the same field twice

File: pkg/sql/colexec/cfetcher.go [hunk 2]
```diff
@@ -69,6 +69,9 @@ type cTableInfo struct {
...
+	// does not contain any -1's. It is meant to be used only in logging.
+	allIndexColOrdinals []int
```
Comment (jordanlewis) [inline pkg/sql/colexec/cfetcher.go]: It seems like we use it for more than just logging, no?

**Comment by rohany** [comment]
This PR actually depends on https://github.com/cockroachdb/cockroach/pull/43002 to pass the tests -- there are tests in master already that test what you're asking for, but this PR removed the code that allowed those tests to pass in favor of combining with https://github.com/cockroachdb/cockroach/pull/43002 to address those issues. I'll let it merge and update the PR.

File: pkg/sql/colexec/cfetcher.go (hunk 2 above)
Comment (rohany) [reply pkg/sql/colexec/cfetcher.go]: we use it to just fill in all of the index columns when printing k/vs to the trace, as opposed to the needed set when tracekv is off

**Comment by rohany** [comment]
RFAL

**Comment by yuzefovich** [comment]
I think this PR fixes #43153, right?

**Comment by rohany** [comment]
I think so? I don't know how to find the schema of the tables in the queries to try it out though...

**Comment by yuzefovich** [comment]
You need to go to `Artifacts` tab of the build that failed. In there, you can find `sqlsmith.log` that contains all the queries that were executed before hitting the error / panic. For example, https://teamcity.cockroachdb.com/viewLog.html?buildId=1643186&buildTypeId=Cockroach_Nightlies_WorkloadNightly&tab=artifacts#%2Fsqlsmith%2Fsetup%3Dempty%2Fsetting%3Ddefault. Also, looking at the logs of the nodes can be helpful to see the stack trace (e.g. https://teamcity.cockroachdb.com/repository/download/Cockroach_Nightlies_WorkloadNightly/1643186:id/sqlsmith/setup%3Dempty/setting%3Ddefault/1.logs/cockroach.log).

**Comment by rohany** [comment]
Yes i see that, but it doesn't show me the initial set of defined tables available to SQLsmith -- the logs just start with the queries it is running.

**Comment by yuzefovich** [comment]
The config is called "empty", so there are no initial tables in that case. If there were, they would also be included in the `sqlsmith.log`.

**Comment by rohany** [comment]
Alright, I've confirmed that https://github.com/cockroachdb/cockroach/issues/43395 is fixed by this commit.

**Review by jordanlewis — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43024)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis)

**Comment by yuzefovich** [comment]
We should backport this commit and unrevert the revert (#43072), right guys? cc @asubiotto

**Comment by rohany** [comment]
This has to be backported too https://github.com/cockroachdb/cockroach/pull/43002

**Comment by asubiotto** [comment]
OK. So to summarize we need to backport #43002, #43024 and then unrevert #43072 in that order?

**Comment by rohany** [comment]
Yes, i think that is correct.

**Pull request merged**